specs/recognition/image_recognition_spec.py

```diff
@@ -0,0 +1,21 @@
+from flask import current_app
+from mamba import description, before, it
+from expects import expect, be_none
+from doublex import Mock, ANY_ARG
+from specs.spec_helpers import Context
+from application.controllers.artifacts_controller import no_content
+from application.recognition.image_recognition import _work_asynchronously
+from specs.factories.image_recognition import vision_api_response
+from unittest.mock import patch
+
+with description('image recognition called during create') as self:
+    with before.each:
+        self.context = Context()
+        with Mock() as artifact_mock:
+            artifact_mock.update(ANY_ARG).returns(no_content())
+            artifact_mock.tags = []
+        self.artifact = artifact_mock
+    with it('Adds Tags to artifact'):
+        with patch('application.recognition.image_recognition._call_google_api',

```

Comment:
you can use the extracted `mock_image_recognition` here as well right?

Reply:
no, this test is intended to test the image_recognition function, so mocking it would achieve nothing.
Instead the _call_google_api function is mocked, so I don't actually do an http call, and instead give back a canned response from the google api, as well as the update function of the artifact, so I don't actually update the database.

I thought I read, that the Mock will raise an error if it is not called during test execution, but if this is not the case, I need to find a better mock.

Reply:
Oh I see that makes more sense. I general I am not a big fan of the patch call. Simply because it doesn't read very well and you have to change it if you move the file. Any specific reason for not making the image_recognition module a class? Then you could simply do:
```
image_recognition = ImageRecognition
Mock(image_recognition._call_google_api).returns(vision_api_response)
image_recognition._work_asynchonously()
```
I think that would be more clear and with doublex-expects, you could also do an expect statement like:
`expect(image_recognition._call_google_api()).to(have_been_called.once)`

It's already installed, documentation can be found here: https://github.com/jaimegildesagredo/doublex-expects

Reply:
This is really hard, spent some hours and I'm not sure it is possible.
It works for artifacts_controller_spec because I don't want the class to do anything, but it fails to work for image_recognition_spec because once I mock the image recognition class, there is no way to make it actually execute any function.
Obviously, I can't test a class, by mocking the class itself.

Reply:
This can be used in some places but does not get rid of the Patch entirely.
I was able to mock the method in image_recognition_spec but in articaft_controller_spec I still need to use patch, since the class is not actually available in scope for mocking.